interfaces/cython/cantera/onedim.py

```diff
@@ -299,6 +300,147 @@ def write_csv(self, filename, species='X', quiet=True):
         if not quiet:
             print("Solution saved to '{0}'.".format(filename))
 
+    def to_solution_array(self):
+        """Return the solution vector as a Cantera SolutionArray object.
+
+        The SolutionArray has the following ``extra`` entries:
+         * `grid`: grid point positions along the flame [m]
+         * `velocity`: normal velocity [m/s]
+         * `gradient`: tangential velocity gradient [1/s] (if applicable)
+         * `lambda`: radial pressure gradient [N/m^4] (if applicable)
+         * `eField`: electric field strength (if applicable)
+        """
+        # create extra columns
+        dom = self.domains[1]
+        extra = {}
+        for e in self._extra:
+            if e == 'grid':
+                extra[e] = self.grid
+            elif e == 'velocity':
+                extra[e] = self.profile(dom, 'u')
+            elif e == 'gradient':
+                extra[e] = self.profile(dom, 'V')
+            else:
+                extra[e] = self.profile(dom, e)
+
+        # create solution array object
+        arr = SolutionArray(self.gas, self.flame.n_points, extra=extra)
+
+        # retrieve species concentrations and set states
+        X = []
+        for n in range(self.flame.n_points):
+            self.set_gas_state(n)
+            X.append(self.gas.X)
+        arr.TPX = self.T, self.gas.P, np.vstack(X)
+
+        return arr
+
+    def from_solution_array(self, arr):
+        """Restore the solution vector from a Cantera SolutionArray object.
+
+        The SolutionArray requires the following ``extra`` entries:
+         * `grid`: grid point positions along the flame [m]
+         * `velocity`: normal velocity [m/s]
+         * `gradient`: tangential velocity gradient [1/s] (if applicable)
+         * `lambda`: radial pressure gradient [N/m^4] (if applicable)
+         * `eField`: electric field strength (if applicable)
+        """
+        # restore grid
+        self.domains[1].grid = arr.grid
+        self._get_initial_solution()
```

Comment:
It actually is necessary as there’s a segfault otherwise (I realize that this used to work several releases ago). I’m likely putting a band-aid here for something that should be fixed elsewhere.

Reply:
Well - I looked into this and the reason is that the `Sim1D` object doesn't know that the vector sizes have been updated for one of its constituting `Domain1D`'s. There are probably ways to circumnavigate this issue, but calling `_get_initial_solution` here is *much* simpler.